nexus/src/app/background/plan_execution.rs

```diff
@@ -0,0 +1,408 @@
+            .sled_id(*sled_id)
+            .fetch()
+            .await
+            .with_context(|| {
+                format!(
+                    "Failed to create sled_agent::Client for sled_id: {}",
+                    sled_id
+                )
+            })?;
+        let dur = std::time::Duration::from_secs(60);
+        let client = reqwest::ClientBuilder::new()
+            .connect_timeout(dur)
+            .timeout(dur)
+            .build()
+            .unwrap();
+        Ok(SledAgentClient::new_with_client(
+            &format!("http://{}", sled.address()),
+            client,
+            opctx.log.clone(),
+        ))
+    }
+
+    async fn realize_blueprint(
+        &self,
+        opctx: &OpContext,
+        blueprint: &std::sync::Arc<Blueprint>,
+    ) -> Result<(), Vec<anyhow::Error>> {
+        let log = opctx.log.new(o!("comment" => blueprint.comment.clone()));
+        self.deploy_zones(&log, opctx, &blueprint.omicron_zones).await
+    }
+
+    async fn deploy_zones(
+        &self,
+        log: &Logger,
+        opctx: &OpContext,
+        zones: &BTreeMap<Uuid, OmicronZonesConfig>,
+    ) -> Result<(), Vec<anyhow::Error>> {
+        let errors: Vec<_> = stream::iter(zones.clone())
+            .filter_map(|(sled_id, config)| async move {
+                let client = match self.sled_client(&opctx, &sled_id).await {
+                    Ok(client) => client,
+                    Err(err) => {
+                        warn!(log, "{}", err);
+                        return Some(err);
+                    }
+                };
+                let result = client
+                    .omicron_zones_put(&config)
+                    .await
+                    .with_context(|| {
+                        format!("Failed to put {config:#?} to sled {sled_id}")
+                    });
+
+                match result {
+                    Err(error) => {
+                        warn!(log, "{}", error);
+                        Some(error)
+                    }
+                    Ok(_) => {
+                        info!(
+                            log,
+                            "Successfully deployed zones for sled agent";
+                            "sled_id" => %sled_id,
+                            "generation" => config.generation.to_string()
+                        );
+                        None
+                    }
+                }
+            })
+            .collect()
+            .await;
+
+        if errors.is_empty() {
+            Ok(())
+        } else {
+            Err(errors)
+        }
+    }
+}
+
+impl BackgroundTask for PlanExecutor {
+    fn activate<'a>(
+        &'a mut self,
+        opctx: &'a OpContext,
+    ) -> BoxFuture<'a, serde_json::Value> {
+        async {
+            // Get the latest blueprint, cloning to prevent holding a read lock
+            // on the watch.
+            let blueprint = self.rx_blueprint.borrow().clone();
```

Comment:
I'm not sure if a `watch::Receiver` is the final version of how the current blueprint gets into this task, but just in general, should this be `borrow_and_update()` so `.changed()` will work if we end up using it?

Reply:
Hmm. I don't think it's strictly necessary, but it may prevent extra unnecessary calls to `activate` of this taks. I was copying the usage in `dns_propagation` which is how I ended up with a call to `borrow`. 

We actually do use `changed` inside the [GenericWatcher](https://github.com/oxidecomputer/omicron/blob/main/nexus/src/app/background/common.rs#L446-L462) which is the only spot we actually wait for the receiver. Then when the value has changed, which the call to `changed` in `GenericWatcher` marks as seen, the task is activated and we just read the latest value. However, as you surmised, if we see a different value and don't mark it seen, then changed will fire again unnecessarily. I'll modify this code to use `borrow_and_update` which seems like the better option.

CC @davepacheco

Reply:
Fixed in 24b34f70f76a5795d51987442f392dc374cc71eb